<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>NORDic_UTILS module &mdash; NORDic v2.4.3 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=f9fa89a7"></script>
        <script src="_static/doctools.js?v=888ff710"></script>
        <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            NORDic
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="install.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="content.html">Statement of need</a></li>
<li class="toctree-l1"><a class="reference internal" href="content.html#usage">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="modules.html">NORDic</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">NORDic</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">NORDic_UTILS module</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/NORDic_UTILS.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="nordic-utils-module">
<h1>NORDic_UTILS module<a class="headerlink" href="#nordic-utils-module" title="Permalink to this heading"></a></h1>
<section id="module-NORDic.UTILS.DISGENET_utils">
<span id="nordic-utils-disgenet-utils-module"></span><h2>NORDic.UTILS.DISGENET_utils module<a class="headerlink" href="#module-NORDic.UTILS.DISGENET_utils" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.DISGENET_utils.get_genes_evidences_from_DISGENET">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.DISGENET_utils.</span></span><span class="sig-name descname"><span class="pre">get_genes_evidences_from_DISGENET</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disease</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CURATED'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunksize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.DISGENET_utils.get_genes_evidences_from_DISGENET" title="Permalink to this definition"></a></dt>
<dd><p>Retrieves the references for the association between each gene in the list and the disease</p>
<p>…</p>
<section id="parameters">
<h3>Parameters<a class="headerlink" href="#parameters" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>gene_list<span class="classifier">Python character string list</span></dt><dd><p>list of associated genes</p>
</dd>
<dt>disease<span class="classifier">Python character string</span></dt><dd><p>Concept ID (CID) from MedGen</p>
</dd>
<dt>limit<span class="classifier">Python integer</span></dt><dd><p>[default=3000] : limit of the number of references</p>
</dd>
<dt>source<span class="classifier">Python character string</span></dt><dd><p>[default=”CURATED”] : DisGeNET data sources [“CURATED”,”ANIMAL MODELS”,”INFERRED”,”ALL”] (see <a class="reference external" href="https://www.disgenet.org/dbinfo">https://www.disgenet.org/dbinfo</a>)</p>
</dd>
<dt>min_score<span class="classifier">Python float</span></dt><dd><p>[default=0] : minimum evidence score</p>
</dd>
<dt>chunksize<span class="classifier">Python integer</span></dt><dd><p>[default=100] : size of chunks (1 chunk per request)</p>
</dd>
<dt>user_key<span class="classifier">Python character string or None</span></dt><dd><p>[default=None] : API key from DisGeNET</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="returns">
<h3>Returns<a class="headerlink" href="#returns" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>res_df<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[row number] x columns/[“gene_symbol”, “sentence”, “associationtype”, “pmid”, “year”, “score”]</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.DISGENET_utils.get_genes_proteins_from_DISGENET">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.DISGENET_utils.</span></span><span class="sig-name descname"><span class="pre">get_genes_proteins_from_DISGENET</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">disease_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'CURATED'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_ei</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dsi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.25</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_dpi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunksize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.DISGENET_utils.get_genes_proteins_from_DISGENET" title="Permalink to this definition"></a></dt>
<dd><p>Retrieves a list of protein names (and associated gene names) related to the input disease CIDs</p>
<p>…</p>
<section id="id1">
<h3>Parameters<a class="headerlink" href="#id1" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>disease_list<span class="classifier">Python character string list</span></dt><dd><p>list of Concept IDs (CID) from Medgen for each disease</p>
</dd>
<dt>limit<span class="classifier">Python integer</span></dt><dd><p>[default=3000] : max. number of proteins</p>
</dd>
<dt>source<span class="classifier">Python character string</span></dt><dd><p>[default=”CURATED”] : DisGeNET data sources [“CURATED”,”ANIMAL MODELS”,”INFERRED”,”ALL”] (see <a class="reference external" href="https://www.disgenet.org/dbinfo">https://www.disgenet.org/dbinfo</a>)</p>
</dd>
<dt>min_score<span class="classifier">Python float</span></dt><dd><p>[default=0] : minimum global score</p>
</dd>
<dt>min_ei<span class="classifier">Python float</span></dt><dd><p>[default=0] : minimimum Evidence Index</p>
</dd>
<dt>min_dsi<span class="classifier">Python float</span></dt><dd><p>[default=0.25] : minimum Disease Specificity Index</p>
</dd>
<dt>min_dpi<span class="classifier">Python float</span></dt><dd><p>[default=0] : minimum Disease Pleiotropy Index</p>
</dd>
<dt>chunksize<span class="classifier">Python integer</span></dt><dd><p>[default=100] : size of chunks (1 chunk per request)</p>
</dd>
<dt>user_key<span class="classifier">Python character string or None</span></dt><dd><p>[default=None] : API key from DisGeNET</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id2">
<h3>Returns<a class="headerlink" href="#id2" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>res_df<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[Disease CID] x columns/[“Protein”, “Gene Name”] or None if Not found.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.DISGENET_utils.get_user_key_DISGENET">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.DISGENET_utils.</span></span><span class="sig-name descname"><span class="pre">get_user_key_DISGENET</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.DISGENET_utils.get_user_key_DISGENET" title="Permalink to this definition"></a></dt>
<dd><p>Retrieves the user key from DisGeNET to call the API</p>
<p>…</p>
<section id="id3">
<h3>Parameters<a class="headerlink" href="#id3" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>path of text file containing on the first line the email, the second the password</p>
</dd>
</dl>
</section>
<section id="id4">
<h3>Returns<a class="headerlink" href="#id4" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>user_key<span class="classifier">Python character string</span></dt><dd><p>from DisGeNET</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
<section id="module-NORDic.UTILS.LINCS_utils">
<span id="nordic-utils-lincs-utils-module"></span><h2>NORDic.UTILS.LINCS_utils module<a class="headerlink" href="#module-NORDic.UTILS.LINCS_utils" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.LINCS_utils.binarize_via_CD">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.LINCS_utils.</span></span><span class="sig-name descname"><span class="pre">binarize_via_CD</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binarize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nperm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10000</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.LINCS_utils.binarize_via_CD" title="Permalink to this definition"></a></dt>
<dd><p>Run a differential expression analysis on a dataframe using Characteristic Direction (CD) [1] (implementation: www.maayanlab.net/CD/)
[1] doi.org/10.1186/1471-2105-15-79</p>
<p>…</p>
<section id="id5">
<h3>Parameters<a class="headerlink" href="#id5" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>df<span class="classifier">Pandas DataFrame</span></dt><dd><p>one transcriptional profile per column (/!if #genes&gt;25,000, then the 25,000 genes with highest variance will be considered)</p>
</dd>
<dt>samples<span class="classifier">Python integer list</span></dt><dd><p>indicates which columns correspond to control (=1) / treated (=2) samples</p>
</dd>
<dt>binarize<span class="classifier">Python integer</span></dt><dd><p>[default=1] : whether to return a binary signature or a real-valued column ~magnitude of change in expression</p>
</dd>
<dt>nperm<span class="classifier">Python integer</span></dt><dd><p>[default=10000] : number of iterations to build the null distribution on which p-values will be computed</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id6">
<h3>Returns<a class="headerlink" href="#id6" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>signature<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[gene index] x columns/[“aggregated”]: 0=down-regulated (DR), 1=up-regulated (UR) (if binarize=1) else &lt;0=DR, &gt;0=UR</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.LINCS_utils.build_url">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.LINCS_utils.</span></span><span class="sig-name descname"><span class="pre">build_url</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">endpoint</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_key</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.LINCS_utils.build_url" title="Permalink to this definition"></a></dt>
<dd><p>Builds the request to CLUE API</p>
<p>…</p>
<section id="id7">
<h3>Parameters<a class="headerlink" href="#id7" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>endpoint<span class="classifier">Python character string</span></dt><dd><p>in [“sigs”, “cells”, “genes”, “perts”, “plates”, “profiles”, “rep_drugs”, “rep_drug_indications”, “pcls”]</p>
</dd>
<dt>method<span class="classifier">Python character string</span></dt><dd><p>in [“count”, “filter”, “distinct”]</p>
</dd>
<dt>params<span class="classifier">Python dictionary</span></dt><dd><p>additional arguments for the request</p>
</dd>
<dt>user_key<span class="classifier">Python character string</span></dt><dd><p>[default=None] : API key for LINCS CLUE.io</p>
</dd>
</dl>
</section>
<section id="id8">
<h3>Returns<a class="headerlink" href="#id8" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>url<span class="classifier">Python character string</span></dt><dd><p>URL of request</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.LINCS_utils.compute_interference_scale">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.LINCS_utils.</span></span><span class="sig-name descname"><span class="pre">compute_interference_scale</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sigs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entrez_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">is_oe</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lincs_specific_ctl_genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">eps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2e-07</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.LINCS_utils.compute_interference_scale" title="Permalink to this definition"></a></dt>
<dd><p>Computes the interference scale [1] which determines whether a genetic perturbation was successful
[1] doi.org/10.1002/psp4.12107</p>
<p>…</p>
<section id="id9">
<h3>Parameters<a class="headerlink" href="#id9" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>sigs<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[control and treated samples]</p>
</dd>
<dt>samples<span class="classifier">Python integer list</span></dt><dd><p>contains 1 for control samples, 2 for treated ones for each column of &#64;sigs</p>
</dd>
<dt>entrez_id<span class="classifier">Python integer</span></dt><dd><p>EntrezID of the perturbed gene</p>
</dd>
<dt>is_oe<span class="classifier">Python bool</span></dt><dd><p>is the experiment an overexpression of the perturbed gene (is_oe=True) or a knockdown</p>
</dd>
<dt>taxon_id<span class="classifier">Python integer</span></dt><dd><p>NCBI taxonomy ID</p>
</dd>
<dt>lincs_specific_ctl_genes<span class="classifier">Python string list</span></dt><dd><p>list of HNCG gene symbols for housekeeping genes</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=True] : prints out verbose</p>
</dd>
<dt>eps<span class="classifier">Python float</span></dt><dd><p>[default=2e-7] : avoids numerical errors for low-expression housekeeping genes</p>
</dd>
</dl>
</section>
<section id="id10">
<h3>Returns<a class="headerlink" href="#id10" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>iscale<span class="classifier">Python float</span></dt><dd><p>interference scale for the input experiment</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.LINCS_utils.convert_ctrlgenes_EntrezGene">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.LINCS_utils.</span></span><span class="sig-name descname"><span class="pre">convert_ctrlgenes_EntrezGene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.LINCS_utils.convert_ctrlgenes_EntrezGene" title="Permalink to this definition"></a></dt>
<dd><p>Retrieves EntrezID from control genes in LINCS L1000 [1]
[1] doi.org/10.1002/psp4.12107</p>
<p>…</p>
<section id="id11">
<h3>Parameters<a class="headerlink" href="#id11" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>taxon_id<span class="classifier">Python integer</span></dt><dd><p>NCBI taxonomy ID</p>
</dd>
</dl>
</section>
<section id="id12">
<h3>Returns<a class="headerlink" href="#id12" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>lincs_specific_ctl_genes<span class="classifier">Python character string list</span></dt><dd><p>list of EntrezGene IDs for all genes in input list</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.LINCS_utils.create_restricted_drug_signatures">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.LINCS_utils.</span></span><span class="sig-name descname"><span class="pre">create_restricted_drug_signatures</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">sig_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entrezid_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_lincs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">which_lvl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[3]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.LINCS_utils.create_restricted_drug_signatures" title="Permalink to this definition"></a></dt>
<dd><p>Create dataframe of drug signatures from LINCS L1000 from a subset of signature and gene IDs</p>
<p>…</p>
<section id="id13">
<h3>Parameters<a class="headerlink" href="#id13" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>sig_ids<span class="classifier">Python character string list</span></dt><dd><p>list of signature IDs from LINCS L1000 (Level 3: “distil_id”, Level 5: “sig_id”)</p>
</dd>
<dt>entrezid_list<span class="classifier">Python character string list</span></dt><dd><p>list of EntrezIDs</p>
</dd>
<dt>path_to_lincs<span class="classifier">Python character string</span></dt><dd><p>folder in which LINCS L1000-related files are stored</p>
</dd>
<dt>which_lvl<span class="classifier">Python integer list</span></dt><dd><p>[3] for Level 3, [5] for Level 5</p>
</dd>
<dt>strict<span class="classifier">Python bool</span></dt><dd><p>[default=True] : if set to True, if not all signatures are retrieved, then return None. If set to False, return the (sub)set of retrievable signatures</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id14">
<h3>Returns<a class="headerlink" href="#id14" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>sigs        Pandas DataFrame</dt><dd><p>rows/[genes] x columns/[drugs]</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.LINCS_utils.download_file">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.LINCS_utils.</span></span><span class="sig-name descname"><span class="pre">download_file</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_to_lincs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">base_url</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_sha</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">check_SHA</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.LINCS_utils.download_file" title="Permalink to this definition"></a></dt>
<dd><p>Downloads automatically LINCS L1000-related files from Gene Expression Omnibus (GEO) (/!can be time-consuming: expect waiting times up to 20 min with a good Internet connection)</p>
<p>…</p>
<section id="id15">
<h3>Parameters<a class="headerlink" href="#id15" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>path_to_lincs<span class="classifier">Python character string</span></dt><dd><p>path to local LINCS L1000 folder in which the files will be downloaded</p>
</dd>
<dt>file_name<span class="classifier">Python character string</span></dt><dd><p>file name to download on GEO</p>
</dd>
<dt>base_url<span class="classifier">Python character string</span></dt><dd><p>path to GEO repository</p>
</dd>
<dt>file_sha<span class="classifier">Python character string</span></dt><dd><p>file name of corresponding SHA hash to check file integrity</p>
</dd>
<dt>check_SHA<span class="classifier">Python bool</span></dt><dd><p>[default=True] : whether to check the file integrity</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id16">
<h3>Returns<a class="headerlink" href="#id16" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>0<span class="classifier">Python integer</span></dt><dd><p>0 means that the download was successful</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.LINCS_utils.download_lincs_files">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.LINCS_utils.</span></span><span class="sig-name descname"><span class="pre">download_lincs_files</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">path_to_lincs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">which_lvl</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.LINCS_utils.download_lincs_files" title="Permalink to this definition"></a></dt>
<dd><p>Returns and downloads the proper LINCS L1000 files from Gene Expression Omnibus (GEO)</p>
<p>…</p>
<section id="id17">
<h3>Parameters<a class="headerlink" href="#id17" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>path_to_lincs<span class="classifier">Python character string</span></dt><dd><p>path to folder in which LINCS L1000-related files will be locally stored</p>
</dd>
<dt>which_lvl<span class="classifier">Python integer list</span></dt><dd><p>LINCS L1000 Level to download (either [3] -normalized gene expression-, [5] -binary experimental signatures-, [3,5])</p>
</dd>
</dl>
</section>
<section id="id18">
<h3>Returns<a class="headerlink" href="#id18" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>file_list<span class="classifier">Python list of 4 Python character string lists</span></dt><dd><p>gene_files, sig_files, lvl3_files, lvl5_files Python lists of character strings</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.LINCS_utils.get_treated_control_dataset">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.LINCS_utils.</span></span><span class="sig-name descname"><span class="pre">get_treated_control_dataset</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">treatment</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pert_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entrez_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_lincs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">entrez_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'distil_ss'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iunit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">itime</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">which_lvl</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[5]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nsigs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">same_plate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trim_w_interference_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_metrics</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.LINCS_utils.get_treated_control_dataset" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve set of experimental profiles, with at least nsigs treated and control sample</p>
<p>…</p>
<section id="id19">
<h3>Parameters<a class="headerlink" href="#id19" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>treatment<span class="classifier">Python character string</span></dt><dd><p>HUGO gene symbol</p>
</dd>
<dt>pert_type<span class="classifier">Python character string</span></dt><dd><p>type of perturbation as accepted by LINCS L1000</p>
</dd>
<dt>cell<span class="classifier">Python character string</span></dt><dd><p>cell line existing in LINCS L1000</p>
</dd>
<dt>filters<span class="classifier">Python dictionary</span></dt><dd><p>additional parameters for the LINCS L1000 requests</p>
</dd>
<dt>entrez_ids<span class="classifier">Python integer list</span></dt><dd><p>EntrezID genes</p>
</dd>
<dt>taxon_id<span class="classifier">Python integer</span></dt><dd><p>NCBI taxonomy ID</p>
</dd>
<dt>user_key<span class="classifier">Python character string</span></dt><dd><p>LINCS L1000 user API key</p>
</dd>
<dt>path_to_lincs<span class="classifier">Python character string</span></dt><dd><p>path where LINCS L1000 files are locally stored</p>
</dd>
<dt>entrez_id<span class="classifier">Python integer</span></dt><dd><p>EntrezID identifier for HUGO gene symbol treatment</p>
</dd>
<dt>selection<span class="classifier">Python character string</span></dt><dd><p>[default=”distil_ss”] : LINCS L1000 metric which is maximized by a given experiment</p>
</dd>
<dt>dose<span class="classifier">Python character string or None</span></dt><dd><p>[default=None] : filter by dose (if not None)</p>
</dd>
<dt>iunit<span class="classifier">Python character string or None</span></dt><dd><p>[default=None] : unit of dose</p>
</dd>
<dt>itime<span class="classifier">Python character string or None</span></dt><dd><p>[default=None] : filter by exposure time (if not None)</p>
</dd>
<dt>which_lvl<span class="classifier">Python integer list</span></dt><dd><p>[default=[3]] : LINCS L1000 data level to consider (either 3 or 5)</p>
</dd>
<dt>nsigs<span class="classifier">Python integer</span></dt><dd><p>[default=2] : minimal number of samples of each condition in each experiment</p>
</dd>
<dt>same_plate<span class="classifier">Python bool</span></dt><dd><p>[default=True] : select samples from the same plate for each experiment and condition</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=True] : prints out verbose</p>
</dd>
<dt>trim_w_interference_scale<span class="classifier">Python bool</span></dt><dd><p>[default=True] : computes the interference scale criteria for further trimming</p>
</dd>
<dt>return_metrics<span class="classifier">Python character string list</span></dt><dd><p>[default=[]] : list of LINCS L1000 metrics to return as the same time as the profiles</p>
</dd>
</dl>
</section>
<section id="id20">
<h3>Returns<a class="headerlink" href="#id20" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>sigs<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes+”annotation”+”signame”+”sigid”] x columns/[profiles] or None</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.LINCS_utils.get_user_key">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.LINCS_utils.</span></span><span class="sig-name descname"><span class="pre">get_user_key</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.LINCS_utils.get_user_key" title="Permalink to this definition"></a></dt>
<dd><p>Retrieves user key for interacting with LINCS L1000 CLUE API</p>
<p>…</p>
<section id="id21">
<h3>Parameters<a class="headerlink" href="#id21" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>path to file containing credentials for LINCS L1000 (first line: username, second line: password, third line: user key)</p>
</dd>
</dl>
</section>
<section id="id22">
<h3>Returns<a class="headerlink" href="#id22" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>user_key<span class="classifier">Python character string</span></dt><dd><p>identifier for the LINCS L1000 CLUE API</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.LINCS_utils.post_request">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.LINCS_utils.</span></span><span class="sig-name descname"><span class="pre">post_request</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">url</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pause_time</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.LINCS_utils.post_request" title="Permalink to this definition"></a></dt>
<dd><p>Post request to API</p>
<p>…</p>
<section id="id23">
<h3>Parameters<a class="headerlink" href="#id23" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>url<span class="classifier">Python character string</span></dt><dd><p>URL formatted as in build_url</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=True] : prints out verbose</p>
</dd>
<dt>pause_time<span class="classifier">Python integer</span></dt><dd><p>[default=1] : minimum time in seconds between each request</p>
</dd>
</dl>
</section>
<section id="id24">
<h3>Returns<a class="headerlink" href="#id24" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>data<span class="classifier">Python dictionary </span></dt><dd><p>(JSON) or Python character string list (if request was method=”distinct”)</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.LINCS_utils.select_best_sig">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.LINCS_utils.</span></span><span class="sig-name descname"><span class="pre">select_best_sig</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filters</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'distil_ss'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nsigs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">same_plate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iunit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.LINCS_utils.select_best_sig" title="Permalink to this definition"></a></dt>
<dd><p>Select “best” set of profiles (“experiment”) (in terms of quality, or criterion “selection”) according to filters</p>
<p>…</p>
<section id="id25">
<h3>Parameters<a class="headerlink" href="#id25" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>params<span class="classifier">Python dictionary</span></dt><dd><p>additional arguments for the request</p>
</dd>
<dt>filters<span class="classifier">Python dictionary</span></dt><dd><p>additional arguments for filtering the results of the request (defined with params)</p>
</dd>
<dt>selection<span class="classifier">Python character string</span></dt><dd><p>[default=”distil_ss”] : name of the metric in LINCS L1000 to define the best signature</p>
</dd>
<dt>nsigs<span class="classifier">Python integer</span></dt><dd><p>[default=2] : minimum number of signatures to retrieve</p>
</dd>
<dt>same_plate<span class="classifier">Python bool</span></dt><dd><p>[default=True] : whether to retrieve signatures from the same plate or not</p>
</dd>
<dt>iunit<span class="classifier">Python character string or None</span></dt><dd><p>[default=None] : unit of dose (if None, any)</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id26">
<h3>Returns<a class="headerlink" href="#id26" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>data<span class="classifier">Python dictionary list</span></dt><dd><p>the list of profile IDs to retrieve from LINCS L1000</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
<section id="module-NORDic.UTILS.STRING_utils">
<span id="nordic-utils-string-utils-module"></span><h2>NORDic.UTILS.STRING_utils module<a class="headerlink" href="#module-NORDic.UTILS.STRING_utils" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.STRING_utils.get_app_name_STRING">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.STRING_utils.</span></span><span class="sig-name descname"><span class="pre">get_app_name_STRING</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.STRING_utils.get_app_name_STRING" title="Permalink to this definition"></a></dt>
<dd><p>Retrieves app name from STRING to interact with the API</p>
<p>…</p>
<section id="id27">
<h3>Parameters<a class="headerlink" href="#id27" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>path to file with a unique line = email adress</p>
</dd>
</dl>
</section>
<section id="id28">
<h3>Returns<a class="headerlink" href="#id28" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>app_name<span class="classifier">Python character string</span></dt><dd><p>identifier for the STRING API</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.STRING_utils.get_image_from_STRING">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.STRING_utils.</span></span><span class="sig-name descname"><span class="pre">get_image_from_STRING</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">my_genes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'network.png'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_flavor</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'evidence'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'functional'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">app_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'11.5'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.STRING_utils.get_image_from_STRING" title="Permalink to this definition"></a></dt>
<dd><p>Retrieves protein IDs in STRING associated with input genes in the correct species</p>
<p>…</p>
<section id="id29">
<h3>Parameters<a class="headerlink" href="#id29" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>genes_list<span class="classifier">Python character list</span></dt><dd><p>list of gene symbols</p>
</dd>
<dt>taxon_id<span class="classifier">Python integer</span></dt><dd><p>taxon ID from NCBI</p>
</dd>
<dt>file_name<span class="classifier">Python character string</span></dt><dd><p>[default=”network.png”] : image file name</p>
</dd>
<dt>min_score<span class="classifier">Python float</span></dt><dd><p>[default=0] : confidence lower threshold (in [0,1])</p>
</dd>
<dt>network_flavor<span class="classifier">Python character string</span></dt><dd><p>[default=”evidence”] : show links related to [“confidence”, “action”, “evidence”]</p>
</dd>
<dt>network_type<span class="classifier">Python character string</span></dt><dd><p>[default=”functional”] : show “functional” or “physical” network</p>
</dd>
<dt>app_name<span class="classifier">Python character string</span></dt><dd><p>[default=None] : identifier for STRING</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id30">
<h3>Returns<a class="headerlink" href="#id30" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>writes the network image to a file file_name</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.STRING_utils.get_interactions_from_STRING">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.STRING_utils.</span></span><span class="sig-name descname"><span class="pre">get_interactions_from_STRING</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">app_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'11.0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.STRING_utils.get_interactions_from_STRING" title="Permalink to this definition"></a></dt>
<dd><p>Retrieves (un)directed and (un)signed physical interactions from the STRING database</p>
<p>…</p>
<section id="id31">
<h3>Parameters<a class="headerlink" href="#id31" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>gene_list<span class="classifier">Python character string list</span></dt><dd><p>list of genes</p>
</dd>
<dt>taxon_id<span class="classifier">Python integer</span></dt><dd><p>NCBI taxonomy ID</p>
</dd>
<dt>min_score<span class="classifier">Python integer</span></dt><dd><p>[default=0] : in [0,1] STRING combined score</p>
</dd>
<dt>app_name<span class="classifier">Python character string</span></dt><dd><p>[default=None] : identifier for STRING</p>
</dd>
<dt>file_folder<span class="classifier">Python character string</span></dt><dd><p>[default=None]: where to save the file from STRING (if None, the file is not saved)</p>
</dd>
<dt>version<span class="classifier">Python character string</span></dt><dd><p>[default=”v11.0”] : STRING database version</p>
</dd>
<dt>strict<span class="classifier">Python bool</span></dt><dd><p>[default=False] : if set to True, only keep interactions involving genes BOTH in &#64;gene_list</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id32">
<h3>Returns<a class="headerlink" href="#id32" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>res_df<span class="classifier">Pandas Dataframe</span></dt><dd><p>rows/[interation number] x columns/[“preferredName_A”, “preferredName_B”, “sign”, “directed”, “score”]</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.STRING_utils.get_interactions_partners_from_STRING">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.STRING_utils.</span></span><span class="sig-name descname"><span class="pre">get_interactions_partners_from_STRING</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'functional'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">app_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'11.5'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.STRING_utils.get_interactions_partners_from_STRING" title="Permalink to this definition"></a></dt>
<dd><p>Retrieves undirected and unsigned interactions from the STRING database</p>
<p>…</p>
<section id="id33">
<h3>Parameters<a class="headerlink" href="#id33" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>gene_list<span class="classifier">Python character string list</span></dt><dd><p>list of gene symbols</p>
</dd>
<dt>taxon_id<span class="classifier">Python integer</span></dt><dd><p>NCBI taxonomy ID</p>
</dd>
<dt>min_score<span class="classifier">Python integer</span></dt><dd><p>[default=0] : minimum STRING combined edge score in [0,1]</p>
</dd>
<dt>network_type<span class="classifier">Python character string</span></dt><dd><p>[default=”functional”] : returns “functional” or “physical” network</p>
</dd>
<dt>limit<span class="classifier">Python integer</span></dt><dd><p>[default=5] : limits the number of interaction partners retrieved per protein (most confident interactions come first)</p>
</dd>
<dt>app_name<span class="classifier">Python character string</span></dt><dd><p>[default=None] : identifier for STRING</p>
</dd>
<dt>version<span class="classifier">Python character string</span></dt><dd><p>[default=”11.5”] : STRING version</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id34">
<h3>Returns<a class="headerlink" href="#id34" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>network<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[row number] x columns/[“preferredName_A”,”preferredName_B”,”score”,”directed”]</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.STRING_utils.get_network_from_STRING">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.STRING_utils.</span></span><span class="sig-name descname"><span class="pre">get_network_from_STRING</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'functional'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">app_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'11.5'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.STRING_utils.get_network_from_STRING" title="Permalink to this definition"></a></dt>
<dd><p>Retrieves undirected and unsigned interactions from the STRING database</p>
<p>…</p>
<section id="id35">
<h3>Parameters<a class="headerlink" href="#id35" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>gene_list<span class="classifier">Python character string list</span></dt><dd><p>list of gene symbols</p>
</dd>
<dt>taxon_id<span class="classifier">Python integer</span></dt><dd><p>NCBI taxonomy ID</p>
</dd>
<dt>min_score<span class="classifier">Python integer</span></dt><dd><p>[default=0] : minimum STRING combined edge score in [0,1]</p>
</dd>
<dt>network_type<span class="classifier">Python character string</span></dt><dd><p>[default=”functional”] : returns “functional” or “physical” network</p>
</dd>
<dt>add_nodes<span class="classifier">Python integer</span></dt><dd><p>[default=0] : add nodes <em>in the closest interaction neighborhood</em> involved with the genes in &#64;gene_list if set to 1</p>
</dd>
<dt>app_name<span class="classifier">Python character string</span></dt><dd><p>[default=None] : identifier for STRING</p>
</dd>
<dt>version<span class="classifier">Python character string</span></dt><dd><p>[default=”11.5”] : STRING version</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id36">
<h3>Returns<a class="headerlink" href="#id36" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>network<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[row number] x columns/[“preferredName_A”,”preferredName_B”,”score”,”directed”]</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.STRING_utils.get_protein_names_from_STRING">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.STRING_utils.</span></span><span class="sig-name descname"><span class="pre">get_protein_names_from_STRING</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">app_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'11.5'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.STRING_utils.get_protein_names_from_STRING" title="Permalink to this definition"></a></dt>
<dd><p>Retrieves protein IDs in STRING associated with input genes in the correct species</p>
<p>…</p>
<section id="id37">
<h3>Parameters<a class="headerlink" href="#id37" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>genes_list<span class="classifier">Python character list</span></dt><dd><p>list of gene symbols</p>
</dd>
<dt>taxon_id<span class="classifier">Python integer</span></dt><dd><p>taxon ID from NCBI</p>
</dd>
<dt>version<span class="classifier">Python character string</span></dt><dd><p>[default=”11.5”] : STRING version</p>
</dd>
<dt>app_name<span class="classifier">Python character string</span></dt><dd><p>[default=None] : identifier to access STRING</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id38">
<h3>Returns<a class="headerlink" href="#id38" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>res_df<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[row number] x columns/[“queryItem”, “stringId”, “preferredName”, “annotation”]</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.STRING_utils.string_api_url">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.STRING_utils.</span></span><span class="sig-name descname"><span class="pre">string_api_url</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">v</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.STRING_utils.string_api_url" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-NORDic.UTILS.utils_data">
<span id="nordic-utils-utils-data-module"></span><h2>NORDic.UTILS.utils_data module<a class="headerlink" href="#module-NORDic.UTILS.utils_data" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_data.convert_EntrezGene_LINCSL1000">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_data.</span></span><span class="sig-name descname"><span class="pre">convert_EntrezGene_LINCSL1000</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_folder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">EntrezGenes</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_data.convert_EntrezGene_LINCSL1000" title="Permalink to this definition"></a></dt>
<dd><p>Converts EntrezIDs to Gene Symbols present in LINCS L1000</p>
<p>…</p>
<section id="id39">
<h3>Parameters<a class="headerlink" href="#id39" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>file_folder<span class="classifier">Python character string</span></dt><dd><p>path to folder of intermediate results</p>
</dd>
<dt>EntrezGenes<span class="classifier">Python character string list</span></dt><dd><p>list of EntrezGene IDs</p>
</dd>
<dt>user_key<span class="classifier">Python character string</span></dt><dd><p>LINCS L1000 user key</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id40">
<h3>Returns<a class="headerlink" href="#id40" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>Pandas<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[EntrezID] x columns/[“Gene Symbol”,”Entrez ID”] (“-” if they do not exist)</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_data.convert_genes_EntrezGene">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_data.</span></span><span class="sig-name descname"><span class="pre">convert_genes_EntrezGene</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">app_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunksize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">100</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">missing_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{'C11ORF74':</span> <span class="pre">'IFTAP',</span> <span class="pre">'ENSP00000451560':</span> <span class="pre">'TPPP2',</span> <span class="pre">'RP11-566K11.2':</span> <span class="pre">'TUBB4'}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_data.convert_genes_EntrezGene" title="Permalink to this definition"></a></dt>
<dd><p>Convert gene symbols into EntrezGene CID</p>
<p>…</p>
<section id="id41">
<h3>Parameters<a class="headerlink" href="#id41" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>gene_list<span class="classifier">Python character string list</span></dt><dd><p>list of genes</p>
</dd>
<dt>taxon_id<span class="classifier">Python character string</span></dt><dd><p>NCBI taxonomy ID</p>
</dd>
<dt>app_name<span class="classifier">Python character string</span></dt><dd><p>STRING identifier</p>
</dd>
<dt>missing_genes<span class="classifier">Python dictionary of character string x character string</span></dt><dd><p>known conversions</p>
</dd>
<dt>chunksize<span class="classifier">Python integer</span></dt><dd><p>[default=100] : 1 chunk per request</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id42">
<h3>Returns<a class="headerlink" href="#id42" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>res_df<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[“InputValue”] x columns/[“Gene ID”/might be separated by “; “] (“-” if they do not exist) or None if no identifier has been found</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_data.get_all_celllines">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_data.</span></span><span class="sig-name descname"><span class="pre">get_all_celllines</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pert_inames</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_data.get_all_celllines" title="Permalink to this definition"></a></dt>
<dd><p>Get all cell lines in which one gene in the input list has been specifically perturbed (genetic perturbation)</p>
<p>…</p>
<section id="id43">
<h3>Parameters<a class="headerlink" href="#id43" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>pert_inames<span class="classifier">Python character string</span></dt><dd><p>List of genes (symbols from LINCS L1000)</p>
</dd>
<dt>user_key<span class="classifier">Python character string</span></dt><dd><p>user key from LINCS L1000 CLUE API</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id44">
<h3>Returns<a class="headerlink" href="#id44" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>cell_lines<span class="classifier">Python character string list</span></dt><dd><p>list of cell lines in which at least one gene from pert_inames has been perturbed</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_data.request_biodbnet">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_data.</span></span><span class="sig-name descname"><span class="pre">request_biodbnet</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">probe_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chunksize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_data.request_biodbnet" title="Permalink to this definition"></a></dt>
<dd><p>Converts gene identifier from from to to in a given species</p>
<p>…</p>
<section id="id45">
<h3>Parameters<a class="headerlink" href="#id45" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>probe_list<span class="classifier">Python character string list</span></dt><dd><p>list of probes to convert (of type <a href="#id133"><span class="problematic" id="id134">from_</span></a>)</p>
</dd>
<dt><a href="#id135"><span class="problematic" id="id136">from_</span></a><span class="classifier">Python character string</span></dt><dd><p>an identifier type as recognized by BioDBnet</p>
</dd>
<dt><a href="#id137"><span class="problematic" id="id138">to_</span></a><span class="classifier">Python character string</span></dt><dd><p>an identifier type as recognized by BioDBnet</p>
</dd>
<dt>taxonId<span class="classifier">Python integer</span></dt><dd><p>NCBI taxonomy ID</p>
</dd>
<dt>chunksize<span class="classifier">Python integer</span></dt><dd><p>[default=500] : 1 chunk per request</p>
</dd>
</dl>
</section>
<section id="id46">
<h3>Returns<a class="headerlink" href="#id46" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>res_df<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[“InputValue”/<a href="#id139"><span class="problematic" id="id140">from_</span></a>] x columns[<a href="#id141"><span class="problematic" id="id142">to_</span></a>] (“-” if the identifier has not been found)</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
<section id="module-NORDic.UTILS.utils_exp">
<span id="nordic-utils-utils-exp-module"></span><h2>NORDic.UTILS.utils_exp module<a class="headerlink" href="#module-NORDic.UTILS.utils_exp" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_exp.get_experimental_constraints">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_exp.</span></span><span class="sig-name descname"><span class="pre">get_experimental_constraints</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_folder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_lines</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pert_types</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pert_di</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selection</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_lincs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thres_iscale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nsigs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_exp.get_experimental_constraints" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve experimental profiles from the provided cell lines, perturbation types, list of genes, in the given species (taxon ID)</p>
<p>…</p>
<section id="id47">
<h3>Parameters<a class="headerlink" href="#id47" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>file_folder<span class="classifier">Python character string</span></dt><dd><p>folder where to store intermediary results</p>
</dd>
<dt>cell_lines<span class="classifier">Python character string list</span></dt><dd><p>cell lines present in LINCS L1000</p>
</dd>
<dt>pert_types<span class="classifier">Python character string list</span></dt><dd><p>types of perturbations as supported by LINCS L1000</p>
</dd>
<dt>pert_di<span class="classifier">Python dictionary</span></dt><dd><p>(keys=Python character string, values=Python integer) associates HUGO gene symbols to their EntrezGene IDs</p>
</dd>
<dt>taxon_id<span class="classifier">Python integer</span></dt><dd><p>NCBI taxonomy ID</p>
</dd>
<dt>selection<span class="classifier">Python character string</span></dt><dd><p>LINCS L1000 metric to maximize</p>
</dd>
<dt>user_key<span class="classifier">Python character string</span></dt><dd><p>LINCS L1000 user API key</p>
</dd>
<dt>path_to_lincs<span class="classifier">Python character string</span></dt><dd><p>path to local LINCS L1000 files</p>
</dd>
<dt>thres_iscale<span class="classifier">Python float or None</span></dt><dd><p>[default=None] : lower threshold on the interference scale which quantifies the success of a genetic experiment</p>
</dd>
<dt>nsigs<span class="classifier">Python integer</span></dt><dd><p>[default=2] : minimal number of profiles per experiment and condition</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id48">
<h3>Returns<a class="headerlink" href="#id48" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>signatures<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes+annotations] x columns/[profile/signature IDs]</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_exp.profiles2signatures">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_exp.</span></span><span class="sig-name descname"><span class="pre">profiles2signatures</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">profiles_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">user_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">path_to_lincs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">backgroundfile</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'distil_ss'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thres</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bin_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'binary'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nbackground_limits</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(4,</span> <span class="pre">30)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_exp.profiles2signatures" title="Permalink to this definition"></a></dt>
<dd><p>Convert experimental profiles into signatures (1 for control samples, 1 for treated ones)</p>
<p>…</p>
<section id="id49">
<h3>Parameters<a class="headerlink" href="#id49" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>profiles_df<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes+annotations] x columns/[samples]</p>
</dd>
<dt>user_key<span class="classifier">Python character string</span></dt><dd><p>LINCS L1000 user API key</p>
</dd>
<dt>path_to_lincs<span class="classifier">Python character string</span></dt><dd><p>path to local LINCS L1000 files</p>
</dd>
<dt>save_fname<span class="classifier">Python character string</span></dt><dd><p>path to save normalized expression profiles per cell line</p>
</dd>
<dt>background_file<span class="classifier">Python bool</span></dt><dd><p>[default=False] : retrieves from LINCS L1000 supplementary expression values if set to True to compute more precise basal gene expression levels</p>
</dd>
<dt>selection<span class="classifier">Python character string</span></dt><dd><p>[default=”distil_ss”] : LINCS L1000 metric to maximize for the “background” data</p>
</dd>
<dt>thres<span class="classifier">Python float</span></dt><dd><p>[default=0.5] : threshold for cutoff normalized gene expression values (in [0,0.5])</p>
</dd>
<dt>bin_method<span class="classifier">Python character string</span></dt><dd><p>[default=”binary”] : binarization approach</p>
</dd>
<dt>nbackground_limits<span class="classifier">Python integer tuple</span></dt><dd><p>[default=(4,30)] : lower and upper bounds on the number of profiles for the background expression data</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id50">
<h3>Returns<a class="headerlink" href="#id50" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>signatures_df       Pandas DataFrame</dt><dd><p>rows/[genes] x columns/[signature ID]</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
<section id="module-NORDic.UTILS.utils_grn">
<span id="nordic-utils-utils-grn-module"></span><h2>NORDic.UTILS.utils_grn module<a class="headerlink" href="#module-NORDic.UTILS.utils_grn" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.CL">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">CL</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">influences</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.CL" title="Permalink to this definition"></a></dt>
<dd><p>Computes the average of node-wise clustering coefficients. The clustering coefficient of a node is the ratio of the degree of the considered node and the maximum possible number of connections such that this node and its current neighbors form a clique</p>
<p>…</p>
<section id="id51">
<h3>Parameters<a class="headerlink" href="#id51" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>influences<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[genes]</p>
</dd>
</dl>
</section>
<section id="id52">
<h3>Returns<a class="headerlink" href="#id52" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>CL<span class="classifier">Python float</span></dt><dd><p>network clustering coefficient</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.Centr">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">Centr</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">influences</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.Centr" title="Permalink to this definition"></a></dt>
<dd><p>Computes the network centralization, which is correlated with the similarity of the network to a graph with a star topology</p>
<p>…</p>
<section id="id53">
<h3>Parameters<a class="headerlink" href="#id53" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>influences<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[genes]</p>
</dd>
</dl>
</section>
<section id="id54">
<h3>Returns<a class="headerlink" href="#id54" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>Centr<span class="classifier">Python float</span></dt><dd><p>network centralization</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.DS">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">DS</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">influences</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.DS" title="Permalink to this definition"></a></dt>
<dd><p>Computes the number of edges over the maximum number of possible connections between the nodes in the network</p>
<p>…</p>
<section id="id55">
<h3>Parameters<a class="headerlink" href="#id55" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>influences<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[genes]</p>
</dd>
</dl>
</section>
<section id="id56">
<h3>Returns<a class="headerlink" href="#id56" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>DS<span class="classifier">Python float</span></dt><dd><p>network density</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.GT">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">GT</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">influences</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.GT" title="Permalink to this definition"></a></dt>
<dd><p>Computes the network heterogeneity, which quantifies the non-uniformity of the node degrees across the network</p>
<p>…</p>
<section id="id57">
<h3>Parameters<a class="headerlink" href="#id57" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>influences<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[genes]</p>
</dd>
</dl>
</section>
<section id="id58">
<h3>Returns<a class="headerlink" href="#id58" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>GT<span class="classifier">Python float</span></dt><dd><p>network heterogeneity</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.build_influences">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">build_influences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tau</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">beta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cor_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'pearson'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">expr_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">accept_nonRNA</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.build_influences" title="Permalink to this definition"></a></dt>
<dd><p>Filters out (and signs of unsigned) edges based on gene expression</p>
<p>…</p>
<section id="id59">
<h3>Parameters<a class="headerlink" href="#id59" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>network_df<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[index] x columns/[[“Input”, “Output”, “SSign”]] interactions</p>
</dd>
<dt>tau<span class="classifier">Python float</span></dt><dd><p>threshold on genepairwise expression correlation</p>
</dd>
<dt>beta<span class="classifier">Python integer</span></dt><dd><p>[default=1] : power applied to the adjacency matrix</p>
</dd>
<dt>cor_method<span class="classifier">Python character string</span></dt><dd><p>[default=”pearson”] : type of correlation</p>
</dd>
<dt>expr_df<span class="classifier">Pandas DataFrame</span></dt><dd><p>[default=None] : rows/[genes] x columns/[samples] gene expression data</p>
</dd>
<dt>accept_nonRNA<span class="classifier">Python bool</span></dt><dd><p>[default=False] : if set to False, ignores gene names which are not present in expr_df</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id60">
<h3>Returns<a class="headerlink" href="#id60" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>influences<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[genes] signed adjacency matrix with only interactions s.t. corr^beta&gt;=tau</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.build_observations">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">build_observations</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">grn</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">signatures</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.build_observations" title="Permalink to this definition"></a></dt>
<dd><p>Implement experimental constraints from perturbation experiments in signatures. Experimental constraints are of the form Initial state masked by single-gene perturbation can lead to a steady attractor state Final</p>
<p>…</p>
<section id="id61">
<h3>Parameters<a class="headerlink" href="#id61" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>grn<span class="classifier">InfluenceGraph (from BoneSiS)</span></dt><dd><p>contains topological constraints</p>
</dd>
<dt>signatures<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[experiment IDs]. Experiment IDs is of the form “&lt;pert. gene&gt;_&lt;pert. type&gt;_&lt;…&gt;_&lt;cell line&gt;” (treated) or “initial_&lt;cell line&gt;” (control)</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id62">
<h3>Returns<a class="headerlink" href="#id62" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>BO<span class="classifier">BoNesis object (from BoneSiS)</span></dt><dd><p>BoNesis object which can be evaluated</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.create_grn">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">create_grn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">influences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exact</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_maxclause</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.create_grn" title="Permalink to this definition"></a></dt>
<dd><p>Create a BoneSiS InfluenceGraph</p>
<p>…</p>
<section id="id63">
<h3>Parameters<a class="headerlink" href="#id63" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>influences<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[genes] of interactions, values in {-1,1,0} -1:negative,1:positive,0:absent</p>
</dd>
<dt>exact<span class="classifier">Python bool</span></dt><dd><p>[default=False] : should all interactions be preserved?</p>
</dd>
<dt>max_maxclause<span class="classifier">Python integer</span></dt><dd><p>[default=3] : upper bound on the number of clauses in DNF form</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id64">
<h3>Returns<a class="headerlink" href="#id64" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>grn<span class="classifier">BoneSiS InfluenceGraph class object</span></dt><dd><p>BoneSiS GRN object</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.desirability">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">desirability</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">f_weight_di</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">A</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">B</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.desirability" title="Permalink to this definition"></a></dt>
<dd><p>Harrington’s desirability function, used by [1]
Converts a list of functions to maximize into a single scalar function to maximize with values in [&#64;A,&#64;B]</p>
<dl class="simple">
<dt>[1] <a class="reference external" href="http://ceur-ws.org/Vol-2488/paper17.pdf">http://ceur-ws.org/Vol-2488/paper17.pdf</a></dt><dd><p><a class="reference external" href="https://cran.r-project.org/web/packages/desirability/vignettes/desirability.pdf">https://cran.r-project.org/web/packages/desirability/vignettes/desirability.pdf</a></p>
</dd>
</dl>
<p>…</p>
<section id="id65">
<h3>Parameters<a class="headerlink" href="#id65" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>x<span class="classifier">datapoint</span></dt><dd><p>any input to functions in f_weight_di</p>
</dd>
<dt>f_weight_di<span class="classifier">Python dictionary</span></dt><dd><p>function with arguments as the same type as x, and associated weight</p>
</dd>
<dt>A<span class="classifier">Python float</span></dt><dd><p>[default=0] : lower bound of the function interval</p>
</dd>
<dt>B<span class="classifier">Python float</span></dt><dd><p>[default=1] : upper bound of the function interval</p>
</dd>
</dl>
</section>
<section id="id66">
<h3>Returns<a class="headerlink" href="#id66" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>des(x)<span class="classifier">Python float</span></dt><dd><p>value of the desirability function at point x</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.general_topological_parameter">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">general_topological_parameter</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">influences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">weights</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.general_topological_parameter" title="Permalink to this definition"></a></dt>
<dd><p>Computes the general topological parameter (GTP) associated with the input network</p>
<p>…</p>
<section id="id67">
<h3>Parameters<a class="headerlink" href="#id67" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>influences<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[genes]</p>
</dd>
<dt>weights<span class="classifier">Python dictionary of (Python character string x Python float)</span></dt><dd><p>all keys must be in [“DS”,”CL”,”Centr”,”GT”]</p>
</dd>
</dl>
</section>
<section id="id68">
<h3>Returns<a class="headerlink" href="#id68" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>score<span class="classifier">Python float</span></dt><dd><p>score using the Harrington’s desirability function</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.get_genes_downstream">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">get_genes_downstream</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.get_genes_downstream" title="Permalink to this definition"></a></dt>
<dd><p>Get the list of genes downstream of a gene in a network</p>
<p>…</p>
<section id="id69">
<h3>Parameters<a class="headerlink" href="#id69" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>network_fname<span class="classifier">Python character string</span></dt><dd><p>path to the .BNET file associated with the network</p>
</dd>
<dt>gene<span class="classifier">Python character string</span></dt><dd><p>gene name in the network</p>
</dd>
<dt>n<span class="classifier">Python integer</span></dt><dd><p>[default=-1] : number of recursions (if&lt;0, recursively get all downstream genes)</p>
</dd>
</dl>
</section>
<section id="id70">
<h3>Returns<a class="headerlink" href="#id70" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>lst_downstream<span class="classifier">Python character string list</span></dt><dd><p>list of nodes downstream of &#64;gene</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.get_genes_interactions_from_PPI">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">get_genes_interactions_from_PPI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ppi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">connected</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filtering</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.get_genes_interactions_from_PPI" title="Permalink to this definition"></a></dt>
<dd><p>Filtering edges to decrease computational cost while preserving network connectivity (if needed)</p>
<p>…</p>
<section id="id71">
<h3>Parameters<a class="headerlink" href="#id71" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>ppi<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[index] x columns[{“preferredName_A”, “preferredName_B”, “sign”, “directed”, “score”]]; sign in {-1,1,2}, directed in {0,1}, score in [0,1]</p>
</dd>
<dt>connected<span class="classifier">Python bool</span></dt><dd><p>[default=True] : if set to True, preserve/enforce connectivity on the final network</p>
</dd>
<dt>score<span class="classifier">Python float</span></dt><dd><p>[default=0] : Lower bound on the edge-associated score</p>
</dd>
<dt>filtering<span class="classifier">Python bool</span></dt><dd><p>[default=True] : Whether to filter out edges by a correlation threshold</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id72">
<h3>Returns<a class="headerlink" href="#id72" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>ppi_accepted<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[index] x columns/[[“Input”, “Output”]]</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.get_genes_most_variable">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">get_genes_most_variable</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">control_profiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">treated_profiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.get_genes_most_variable" title="Permalink to this definition"></a></dt>
<dd><p>Get the list of genes which contribute most to the variation between two conditions (in the &#64;pth percentile of change)</p>
<p>…</p>
<section id="id73">
<h3>Parameters<a class="headerlink" href="#id73" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>control_profiles<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[samples] profiles from condition 1</p>
</dd>
<dt>treated_profiles<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[samples] profiles from condition 1</p>
</dd>
<dt>p<span class="classifier">Python float</span></dt><dd><p>100*p th percentile to consider</p>
</dd>
</dl>
</section>
<section id="id74">
<h3>Returns<a class="headerlink" href="#id74" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>lst_genes<span class="classifier">Python character string list</span></dt><dd><p>list of nodes which contribute most to the variation between conditions</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.get_grfs_from_solution">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">get_grfs_from_solution</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">solution</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.get_grfs_from_solution" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve all gene regulatory functions (GRFs) from a given solution</p>
<p>…</p>
<section id="id75">
<h3>Parameters<a class="headerlink" href="#id75" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>solution<span class="classifier">Pandas Series</span></dt><dd><p>rows/[genes]</p>
</dd>
</dl>
</section>
<section id="id76">
<h3>Returns<a class="headerlink" href="#id76" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>grfs<span class="classifier">Python dictionary</span></dt><dd><p>{gene: {regulator: sign, …}, …} where sign in {-1,1} -1: inhibitor, 1: activator</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.get_maxdegree">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">get_maxdegree</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">influences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">activatory</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.get_maxdegree" title="Permalink to this definition"></a></dt>
<dd><p>Computes the maximum ingoing degree (or the maximum number of potential activatory regulators) in a graph</p>
<p>…</p>
<section id="id77">
<h3>Parameters<a class="headerlink" href="#id77" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>influences<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[genes] of interactions: -1:negative,1:positive,0:absent</p>
</dd>
<dt>activatory<span class="classifier">Python bool</span></dt><dd><p>[default=True] : computes the maximum number of potential activatory regulators instead</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id78">
<h3>Returns<a class="headerlink" href="#id78" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>maxindegree<span class="classifier">Python integer</span></dt><dd><p>maximum ingoing degree (or the maximum number of potential activatory regulators)</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.get_minimal_edges">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">get_minimal_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">R</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maximal</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.get_minimal_edges" title="Permalink to this definition"></a></dt>
<dd><p>Return one of the solutions with the smallest (or greatest) number of edges</p>
<p>…</p>
<section id="id79">
<h3>Parameters<a class="headerlink" href="#id79" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>R<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[solution IDs]</p>
</dd>
<dt>connected<span class="classifier">Python bool</span></dt><dd><p>[default=False] : if set to True, return the CONNECTED solution which satisfies those constraints</p>
</dd>
<dt>maximal<span class="classifier">Python bool</span></dt><dd><p>[default=False] : if set to True, return the solution with the greatest number of edges</p>
</dd>
</dl>
</section>
<section id="id80">
<h3>Returns<a class="headerlink" href="#id80" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>solution, nedges<span class="classifier">Python integer x Python integer</span></dt><dd><p>solution and corresponding number of edges</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.get_weakly_connected">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">get_weakly_connected</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">index_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'preferredName_A'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">column_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'preferredName_B'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sscore'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.get_weakly_connected" title="Permalink to this definition"></a></dt>
<dd><p>Depth-first search (DFS) on undirected network</p>
<p>…</p>
<section id="id81">
<h3>Parameters<a class="headerlink" href="#id81" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>network_df<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[index] x columns/[[“Input”,”Output”]]</p>
</dd>
<dt>gene_list<span class="classifier">Python character string list</span></dt><dd><p>list of genes (needed to take into account isolated genes in the network)</p>
</dd>
<dt>index_col<span class="classifier">Python character string</span></dt><dd><p>[default=”preferredName_A”] : column in network_df (input gene)</p>
</dd>
<dt>column_col<span class="classifier">Python character string</span></dt><dd><p>[default=”preferredName_B”] : column in network_df (output gene)</p>
</dd>
<dt>score_col<span class="classifier">Python character string</span></dt><dd><p>[default=”sscore”] : column in network_df (edge weight)</p>
</dd>
</dl>
</section>
<section id="id82">
<h3>Returns<a class="headerlink" href="#id82" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>components<span class="classifier">Type of &#64;network_df.loc[network_df.index[0]][“Input”] Python list of Python list</span></dt><dd><p>list of weakly connected components in the network, ordered by decreasing size</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.infer_network">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">infer_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">BO</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'solutions'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_diverse</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">niterations</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.infer_network" title="Permalink to this definition"></a></dt>
<dd><p>Infer solutions matching topological &amp; experimental constraints</p>
<p>…</p>
<section id="id83">
<h3>Parameters<a class="headerlink" href="#id83" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>BO<span class="classifier">Bonesis object (from BoneSiS)</span></dt><dd><p>contains topological &amp; experimental constraints</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>[default=”solutions”] : path to solution files</p>
</dd>
<dt>use_diverse<span class="classifier">Python bool</span></dt><dd><p>[default=True] : use the “diverse” procedure in BoneSiS</p>
</dd>
<dt>limit<span class="classifier">Python integer</span></dt><dd><p>[default=50] : maximum number of solutions to generate per interation</p>
</dd>
<dt>niterations<span class="classifier">Python integer</span></dt><dd><p>[default=1] : maximum number of iterations</p>
</dd>
</dl>
</section>
<section id="id84">
<h3>Returns<a class="headerlink" href="#id84" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>nsolutions<span class="classifier">Python integer</span></dt><dd><p>list of # solutions per iteration</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.load_grn">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">load_grn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.load_grn" title="Permalink to this definition"></a></dt>
<dd><p>Loads GRN as MPBN class element</p>
<p>…</p>
<section id="id85">
<h3>Parameters<a class="headerlink" href="#id85" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>BNET file</p>
</dd>
</dl>
</section>
<section id="id86">
<h3>Returns<a class="headerlink" href="#id86" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>BN<span class="classifier">mpbn.MPBooleanNetwork object</span></dt><dd><p>Boolean network with Most Permissive semantics</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.reconnect_network">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">reconnect_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.reconnect_network" title="Permalink to this definition"></a></dt>
<dd><p>Write the network with all isolated nodes (no ingoing/outgoing edges) filtered out</p>
<p>…</p>
<section id="id87">
<h3>Parameters<a class="headerlink" href="#id87" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>network_fname<span class="classifier">Python character string</span></dt><dd><p>path to the .BNET associated with the network</p>
</dd>
</dl>
</section>
<section id="id88">
<h3>Returns<a class="headerlink" href="#id88" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>path to the .BNET associated with the reconnected network</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.save_grn">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">save_grn</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">solution</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sep</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">',</span> <span class="pre">'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.save_grn" title="Permalink to this definition"></a></dt>
<dd><p>Write and/or print .bnet file</p>
<p>…</p>
<section id="id89">
<h3>Parameters<a class="headerlink" href="#id89" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>solution<span class="classifier">Pandas Series</span></dt><dd><p>rows/[genes] contains gene regulatory functions (GRF)</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>where to write the file (w/o .bnet extension)</p>
</dd>
<dt>sep<span class="classifier">Python character string</span></dt><dd><p>what separates regulators from regulated genes</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
<dt>max_show<span class="classifier">Python integer</span></dt><dd><p>[default=5] : maximum number of printed GRFs</p>
</dd>
<dt>write<span class="classifier">Python bool</span></dt><dd><p>[default=True] : if set to True, write to a .bnet file</p>
</dd>
</dl>
</section>
<section id="id90">
<h3>Returns<a class="headerlink" href="#id90" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>writes the GRN to a file fname</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.save_solutions">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">save_solutions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">bnetworks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">limit</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.save_solutions" title="Permalink to this definition"></a></dt>
<dd><p>Enumerate and save solutions</p>
<p>…</p>
<section id="id91">
<h3>Parameters<a class="headerlink" href="#id91" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>bnetworks<span class="classifier">Bonesis object</span></dt><dd><p>Output of the inference</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>ZIP filename to store the solutions</p>
</dd>
<dt>limit<span class="classifier">Python integer</span></dt><dd><p>maximum number of solutions to enumerate</p>
</dd>
</dl>
</section>
<section id="id92">
<h3>Parameters<a class="headerlink" href="#id92" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>n<span class="classifier">Python integer</span></dt><dd><p>number of enumerated solutions</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.solution2influences">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">solution2influences</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">solution</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.solution2influences" title="Permalink to this definition"></a></dt>
<dd><p>Converts a solution object into a influences object</p>
<p>…</p>
<section id="id93">
<h3>Parameters<a class="headerlink" href="#id93" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>solution<span class="classifier">Pandas Series</span></dt><dd><p>rows/[genes]</p>
</dd>
</dl>
</section>
<section id="id94">
<h3>Returns<a class="headerlink" href="#id94" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>influences<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[genes] contains values {-1,1,0,2} -1: negative, 1: positive, 0: absent, 2: non monotonic</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_grn.zip2df">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_grn.</span></span><span class="sig-name descname"><span class="pre">zip2df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_grn.zip2df" title="Permalink to this definition"></a></dt>
<dd><p>Extract solutions in ZIP file as DataFrames</p>
<p>…</p>
<section id="id95">
<h3>Parameters<a class="headerlink" href="#id95" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>zip file which contains BNET solutions</p>
</dd>
</dl>
</section>
<section id="id96">
<h3>Returns<a class="headerlink" href="#id96" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>solutions<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[solutions] the GRFs for each gene in each solution</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
<section id="module-NORDic.UTILS.utils_network">
<span id="nordic-utils-utils-network-module"></span><h2>NORDic.UTILS.utils_network module<a class="headerlink" href="#module-NORDic.UTILS.utils_network" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_network.aggregate_networks">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_network.</span></span><span class="sig-name descname"><span class="pre">aggregate_networks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_folder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">taxon_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_score</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">app_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version_net</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'11.5'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version_act</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'11.0'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_network.aggregate_networks" title="Permalink to this definition"></a></dt>
<dd><p>This function performs the following pipeline to build a prior knowledge network based on a subset of genes
- Retrieve protein actions and predicted PPIs from STRING
- Merge the two networks while solving all inconsistencies (duplicates, paradoxes, etc.) in signs, directions, scores
- Determine the greatest threshold on the edge score which allows all of the core gene set to be connected (binary search)
- Trim out edges which scores are below the threshold, and remove all isolated nodes</p>
<p>…</p>
<section id="id97">
<h3>Parameters<a class="headerlink" href="#id97" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>file_folder<span class="classifier">Python character string</span></dt><dd><p>relative path where to store files</p>
</dd>
<dt>gene_list<span class="classifier">Python character string list</span></dt><dd><p>list of core gene symbols to preserve in the network</p>
</dd>
<dt>taxon_id<span class="classifier">Python integer</span></dt><dd><p>NCBI taxonomy ID</p>
</dd>
<dt>min_score<span class="classifier">Python integer</span></dt><dd><p>minimum score on edges retrieved from the STRING database</p>
</dd>
<dt>app_name<span class="classifier">Python character string</span></dt><dd><p>Identifier for STRING requests</p>
</dd>
<dt>version_net<span class="classifier">Python character string</span></dt><dd><p>[default=”11.5”] : Number of version for interaction data in the STRING database. To avoid compatibility issues, it is strongly advised not to change this parameter</p>
</dd>
<dt>version_act<span class="classifier">Python character string</span></dt><dd><p>[default=”11.0”] : Number of version for protein action data in the STRING database. To avoid compatibility issues, it is strongly advised not to change this parameter</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=None] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id98">
<h3>Returns<a class="headerlink" href="#id98" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>final_network<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[interactions] x columns/[[“preferredName_A”, “preferredName_B”, “sign”, “directed”, “score”]]</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_network.capture">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_network.</span></span><span class="sig-name descname"><span class="pre">capture</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_network.capture" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_network.determine_edge_threshold">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_network.</span></span><span class="sig-name descname"><span class="pre">determine_edge_threshold</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">core_gene_set</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_network.determine_edge_threshold" title="Permalink to this definition"></a></dt>
<dd><p>Determine the greatest threshold on the edge score which allows all of the core gene set to be connected (binary search)</p>
<p>…</p>
<section id="id99">
<h3>Parameters<a class="headerlink" href="#id99" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>network<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[interactions] x at least three columns “preferredName_A” (input node), “preferredName_B” (output node), “score” (edge score)</p>
</dd>
<dt>core_gene_set<span class="classifier">Python character string list</span></dt><dd><p>list of genes that should remain connected</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=None]: prints out verbose</p>
</dd>
</dl>
</section>
<section id="id100">
<h3>Returns<a class="headerlink" href="#id100" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>t<span class="classifier">Python float</span></dt><dd><p>maximum threshold which allows the connection of all genes in the core set</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_network.get_network_from_OmniPath">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_network.</span></span><span class="sig-name descname"><span class="pre">get_network_from_OmniPath</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">gene_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">disease_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">species</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'human'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sources_int</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'omnipath'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">domains_int</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">types_int</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_curation_effort</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">domains_annot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'HPA_tissue'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_network.get_network_from_OmniPath" title="Permalink to this definition"></a></dt>
<dd><p>Retrieve a network from OmniPath</p>
<p>…</p>
<section id="id101">
<h3>Parameters<a class="headerlink" href="#id101" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>gene_list<span class="classifier">Python character string</span></dt><dd><p>[default=None] : List of genes to consider (or do not filter the interactions from Omnipath if =None)</p>
</dd>
<dt>disease_name<span class="classifier">Python character string</span></dt><dd><p>[default=None] : Disease name (in letters) to consider</p>
</dd>
<dt>species<span class="classifier">Python character string</span></dt><dd><p>[default=None] : Species to consider (either “human”, “mouse”, or “rat”)</p>
</dd>
<dt>sources_int<span class="classifier">Python character string</span></dt><dd><p>[default=None] : Which databases for interactions to consider (if =None, consider them all)</p>
</dd>
<dt>domains_int<span class="classifier">Python character string</span></dt><dd><p>[default=None] : source of interactions in OmniPath</p>
</dd>
<dt>types_int<span class="classifier">Python character string</span></dt><dd><p>[default=None] : Types of interactions, e.g., “post_translational”, “transcriptional”, “post_transcriptional”, “mirna_transcriptional”</p>
</dd>
<dt>min_curation_effort<span class="classifier">Python integer</span></dt><dd><p>[default=-1] : if positive, select edges based on that criteria (the higher, the better). Counts the unique database-citation pairs, i.e. how many times was an interaction described in a paper and mentioned in a database</p>
</dd>
<dt>domain_annot<span class="classifier">Python character string</span></dt><dd><p>[default=’HPA_tissue’] : source of annotations in OmniPath</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=False] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id102">
<h3>Returns<a class="headerlink" href="#id102" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>final_network<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[interactions] x columns/[[“preferredName_A”, “preferredName_B”, “sign”, “directed”, “score”]]</p>
</dd>
<dt>annot_wide<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[gene symbols] x columns/[annotations from the database &#64;domains_annot]</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_network.merge_network_PPI">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_network.</span></span><span class="sig-name descname"><span class="pre">merge_network_PPI</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">PPI</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_network.merge_network_PPI" title="Permalink to this definition"></a></dt>
<dd><p>Merge two network while solving all inconsistencies (duplicates, paradoxes, etc.) in signs, directions, scores</p>
<p>…</p>
<section id="id103">
<h3>Parameters<a class="headerlink" href="#id103" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>network<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[interactions] x at least three columns “preferredName_A” (input node), “preferredName_B” (output node), “score” (edge score)</p>
</dd>
<dt>PPI<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[interactions] x at least three columns “preferredName_A” (input node), “preferredName_B” (output node), “score” (edge score)</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=None] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id104">
<h3>Returns<a class="headerlink" href="#id104" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>final_network<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[interactions] x columns/[[“preferredName_A”, “preferredName_B”, “sign”, “directed”, “score”]]</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_network.remove_isolated">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_network.</span></span><span class="sig-name descname"><span class="pre">remove_isolated</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">quiet</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_network.remove_isolated" title="Permalink to this definition"></a></dt>
<dd><p>Remove all nodes which do not belong to the largest connected component from the network</p>
<p>…</p>
<section id="id105">
<h3>Parameters<a class="headerlink" href="#id105" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>network<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[interactions] x columns/[[“preferredName_A”, “preferredName_B”, “sign”, “directed”, “score”]]</p>
</dd>
<dt>quiet<span class="classifier">Python bool</span></dt><dd><p>[default=None] : prints out verbose</p>
</dd>
</dl>
</section>
<section id="id106">
<h3>Returns<a class="headerlink" href="#id106" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>trimmed_network<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[interactions] x columns/[[“preferredName_A”, “preferredName_B”, “sign”, “directed”, “score”]]</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
<section id="module-NORDic.UTILS.utils_plot">
<span id="nordic-utils-utils-plot-module"></span><h2>NORDic.UTILS.utils_plot module<a class="headerlink" href="#module-NORDic.UTILS.utils_plot" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_plot.influences2graph">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_plot.</span></span><span class="sig-name descname"><span class="pre">influences2graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">influences</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optional</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compile2png</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">engine</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'sfdp'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_plot.influences2graph" title="Permalink to this definition"></a></dt>
<dd><p>Plots a network by conversion to a DOT file and then to PNG</p>
<p>…</p>
<section id="id107">
<h3>Parameters<a class="headerlink" href="#id107" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>influences<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[genes], contains {-1,1,2}</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>filename of png file</p>
</dd>
<dt>optional<span class="classifier">Python bool</span></dt><dd><p>[default=False] : should interactions be drawn as optional (dashed lines)?</p>
</dd>
</dl>
</section>
<section id="id108">
<h3>Returns<a class="headerlink" href="#id108" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>writes a DOT file which can be converted to PNG image (if compile2png=True)</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_plot.plot_boxplots">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_plot.</span></span><span class="sig-name descname"><span class="pre">plot_boxplots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">scores</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">patient_scores</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ground_truth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">12</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">msize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'boxplots.pdf'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_plot.plot_boxplots" title="Permalink to this definition"></a></dt>
<dd><p>Plots one boxplot per treatment (all values obtained on patient profiles)</p>
<p>…</p>
<section id="id109">
<h3>Parameters<a class="headerlink" href="#id109" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>scores<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[drug names] x column/[value]</p>
</dd>
<dt>patient_scores<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[drug names] x columns/[patient samples]</p>
</dd>
<dt>ground_truth<span class="classifier">Pandas DataFrame</span></dt><dd><p>[default=None] : rows/[drug names] x column/[class] Values in 1: treatment, 0: unknown, -1: aggravating. If not provided: does not color boxplots according to the class</p>
</dd>
<dt>fsize<span class="classifier">Python integer</span></dt><dd><p>[default=18] : font size</p>
</dd>
<dt>msize<span class="classifier">Python integer</span></dt><dd><p>[default=5] : marker size</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>[default=”boxplots”] : file name for the plot</p>
</dd>
</dl>
</section>
<section id="id110">
<h3>Returns<a class="headerlink" href="#id110" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>create boxplots of reward scores across patients for each drug</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_plot.plot_discrete_distributions">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_plot.</span></span><span class="sig-name descname"><span class="pre">plot_discrete_distributions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">signatures</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'signature_expression_distribution.png'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_plot.plot_discrete_distributions" title="Permalink to this definition"></a></dt>
<dd><p>Plots the distributions (histograms) of genes with determined status across signatures</p>
<p>…</p>
<section id="id111">
<h3>Parameters<a class="headerlink" href="#id111" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>signatures<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[samples] with values in {0,NaN,1}. Determined status is either 0 or 1.</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>[default=”signature_expression_distribution.png”] : file name</p>
</dd>
</dl>
</section>
<section id="id112">
<h3>Returns<a class="headerlink" href="#id112" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>plots the number of genes with expression values 0, 1 or NaN in each signature</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_plot.plot_distributions">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_plot.</span></span><span class="sig-name descname"><span class="pre">plot_distributions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">profiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'gene_expression_distribution.png'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thres</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_plot.plot_distributions" title="Permalink to this definition"></a></dt>
<dd><p>Plots the distributions (boxplots) of gene expression across samples for each gene, and the selected threshold for binarization</p>
<p>…</p>
<section id="id113">
<h3>Parameters<a class="headerlink" href="#id113" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>profiles<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes+annotations] x columns/[samples]</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>[default=”gene_expression_distribution.png”] : file name</p>
</dd>
<dt>thres<span class="classifier">Python float or None</span></dt><dd><p>[default=None] : binarization threshold (if there is any)</p>
</dd>
</dl>
</section>
<section id="id114">
<h3>Returns<a class="headerlink" href="#id114" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>plots boxplots of expression for each gene in profiles</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_plot.plot_heatmap">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_plot.</span></span><span class="sig-name descname"><span class="pre">plot_heatmap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">X</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ground_truth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'heatmap.pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">w</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">h</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bfsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">75</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_plot.plot_heatmap" title="Permalink to this definition"></a></dt>
<dd><p>Plots an heatmap of the signatures, with the potential ground truth</p>
<p>…</p>
<section id="id115">
<h3>Parameters<a class="headerlink" href="#id115" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>X<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[features] x columns/[drug names]</p>
</dd>
<dt>ground_truth<span class="classifier">Pandas DataFrame</span></dt><dd><p>[default=None] : rows/[drug names] x column/[class] Values in 1: treatment, 0: unknown, -1: aggravating. If not provided: does not color boxplots according to the class</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>[default=”heatmap.pdf”] : file name for the plot</p>
</dd>
<dt>w<span class="classifier">Python integer</span></dt><dd><p>[default=20] : figure width</p>
</dd>
<dt>h<span class="classifier">Python integer</span></dt><dd><p>[default=20] : figure height</p>
</dd>
<dt>bfsize<span class="classifier">Python integer</span></dt><dd><p>[default=20] : font size in the color bar</p>
</dd>
<dt>rot<span class="classifier">Python integer</span></dt><dd><p>[default=75] : rotation angle of labels</p>
</dd>
</dl>
</section>
<section id="id116">
<h3>Returns<a class="headerlink" href="#id116" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>plots an heatmap of similarity across drugs based on the Pearson correlation</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_plot.plot_influence_graph">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_plot.</span></span><span class="sig-name descname"><span class="pre">plot_influence_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sign_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction_col</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'graph.png'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">optional</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_plot.plot_influence_graph" title="Permalink to this definition"></a></dt>
<dd><p>Converts a network into a PNG picture</p>
<p>…</p>
<section id="id117">
<h3>Parameters<a class="headerlink" href="#id117" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>network_df<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[index] x columns/[input_col,output_col,sign_col]</p>
</dd>
<dt>input_col,output_col,sign_col,direction_col<span class="classifier">Python character string</span></dt><dd><p>columns of network_df</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>[default=”graph.png”] : file name for PNG picture</p>
</dd>
<dt>optional<span class="classifier">Python bool</span></dt><dd><p>[default=True] : should edges be plotted as dashed lines?</p>
</dd>
</dl>
</section>
<section id="id118">
<h3>Returns<a class="headerlink" href="#id118" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>Creates a image of the graph in file fname</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_plot.plot_precision_recall">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_plot.</span></span><span class="sig-name descname"><span class="pre">plot_precision_recall</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">beta</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thres</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'PRC.pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'predictor'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">18</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_plot.plot_precision_recall" title="Permalink to this definition"></a></dt>
<dd><p>Plots a Precision-Recall curve (with variations across samples)</p>
<p>…</p>
<section id="id119">
<h3>Parameters<a class="headerlink" href="#id119" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>pr<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[drug names] x column/[value]</p>
</dd>
<dt>prs<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[drug names] x columns/[patient samples]</p>
</dd>
<dt>tr<span class="classifier">Pandas DataFrame</span></dt><dd><p>[default=None] : rows/[drug names] x column/[class]</p>
</dd>
<dt>beta<span class="classifier">Python float</span></dt><dd><p>[default=1] : value of coefficient beta for the F-measure</p>
</dd>
<dt>thres<span class="classifier">Python float</span></dt><dd><p>[default=0.5] : decision threshold</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>[default=”PRC.pdf”] : file name for the plot</p>
</dd>
<dt>method_name<span class="classifier">Python character string</span></dt><dd><p>[default=”predictor”] : name of the predictor</p>
</dd>
<dt>fsize<span class="classifier">Python integer</span></dt><dd><p>[default=18] : font size</p>
</dd>
</dl>
</section>
<section id="id120">
<h3>Returns<a class="headerlink" href="#id120" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>Plots a Precision-Recall curve based on the drug repurposing predictions</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_plot.plot_roc_curve">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_plot.</span></span><span class="sig-name descname"><span class="pre">plot_roc_curve</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">pr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tr</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ROC.pdf'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'predictor'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fsize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">18</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_plot.plot_roc_curve" title="Permalink to this definition"></a></dt>
<dd><p>Plots a ROC curve (with variations across samples)</p>
<p>…</p>
<section id="id121">
<h3>Parameters<a class="headerlink" href="#id121" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>pr<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[drug names] x column/[value]</p>
</dd>
<dt>prs<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[drug names] x columns/[patient samples]</p>
</dd>
<dt>tr<span class="classifier">Pandas DataFrame</span></dt><dd><p>[default=None] : rows/[drug names] x column/[class]</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>[default=”ROC.pdf”] : file name for the plot</p>
</dd>
<dt>method_name<span class="classifier">Python character string</span></dt><dd><p>[default=”predictor”] : name of the predictor</p>
</dd>
<dt>fsize<span class="classifier">Python integer</span></dt><dd><p>[default=18] : font size</p>
</dd>
</dl>
</section>
<section id="id122">
<h3>Returns<a class="headerlink" href="#id122" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>Plots a ROC curve based on the drug repurposing predictions</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_plot.plot_signatures">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_plot.</span></span><span class="sig-name descname"><span class="pre">plot_signatures</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">signatures</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">perturbed_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">height</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_show</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">50</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'signatures'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_plot.plot_signatures" title="Permalink to this definition"></a></dt>
<dd><p>Print signatures</p>
<p>…</p>
<section id="id123">
<h3>Parameters<a class="headerlink" href="#id123" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>signatures<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[signature IDs]</p>
</dd>
<dt>perturbed_genes<span class="classifier">Python character string list</span></dt><dd><p>[default=None] : list of gene names perturbed in the signatures</p>
</dd>
<dt>width, height<span class="classifier">Python integer</span></dt><dd><p>[default=10] : dimensions of image</p>
</dd>
<dt>max_show<span class="classifier">Python integer</span></dt><dd><p>[default=50] : maximum number of genes shown (as only the &#64;max_show genes with highest variance across signatures are plotted)</p>
</dd>
<dt>fname<span class="classifier">Python character string</span></dt><dd><p>[default=”signatures”] : path of resulting PNG image</p>
</dd>
</dl>
</section>
<section id="id124">
<h3>Returns<a class="headerlink" href="#id124" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>None</dt><dd><p>plots the signatures as heatmaps in file fname</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
<section id="module-NORDic.UTILS.utils_sim">
<span id="nordic-utils-utils-sim-module"></span><h2>NORDic.UTILS.utils_sim module<a class="headerlink" href="#module-NORDic.UTILS.utils_sim" title="Permalink to this heading"></a></h2>
<dl class="py class">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BN_SIM">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_sim.</span></span><span class="sig-name descname"><span class="pre">BN_SIM</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seednb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BN_SIM" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BN_SIM.add_initial_states">
<span class="sig-name descname"><span class="pre">add_initial_states</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">initial</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BN_SIM.add_initial_states" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BN_SIM.add_permanent_mutation">
<span class="sig-name descname"><span class="pre">add_permanent_mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BN_SIM.add_permanent_mutation" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BN_SIM.add_transient_mutation">
<span class="sig-name descname"><span class="pre">add_transient_mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BN_SIM.add_transient_mutation" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BN_SIM.attrs_similarity">
<span class="sig-name descname"><span class="pre">attrs_similarity</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">attrs1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">attrs2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_outputs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BN_SIM.attrs_similarity" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BN_SIM.boxplot">
<span class="sig-name descname"><span class="pre">boxplot</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BN_SIM.boxplot" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BN_SIM.enumerate_attractors">
<span class="sig-name descname"><span class="pre">enumerate_attractors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BN_SIM.enumerate_attractors" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BN_SIM.generate_trajectories">
<span class="sig-name descname"><span class="pre">generate_trajectories</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outputs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BN_SIM.generate_trajectories" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BN_SIM.initialize_network">
<span class="sig-name descname"><span class="pre">initialize_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BN_SIM.initialize_network" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BN_SIM.up_to_attractors">
<span class="sig-name descname"><span class="pre">up_to_attractors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_permanent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_transient</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BN_SIM.up_to_attractors" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BN_SIM.update_network">
<span class="sig-name descname"><span class="pre">update_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">initial</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_permanent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_transient</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BN_SIM.update_network" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BONESIS_SIM">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_sim.</span></span><span class="sig-name descname"><span class="pre">BONESIS_SIM</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seednb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BONESIS_SIM" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#NORDic.UTILS.utils_sim.BN_SIM" title="NORDic.UTILS.utils_sim.BN_SIM"><code class="xref py py-class docutils literal notranslate"><span class="pre">BN_SIM</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BONESIS_SIM.add_initial_states">
<span class="sig-name descname"><span class="pre">add_initial_states</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">initial</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BONESIS_SIM.add_initial_states" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BONESIS_SIM.add_permanent_mutation">
<span class="sig-name descname"><span class="pre">add_permanent_mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BONESIS_SIM.add_permanent_mutation" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BONESIS_SIM.add_transient_mutation">
<span class="sig-name descname"><span class="pre">add_transient_mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BONESIS_SIM.add_transient_mutation" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BONESIS_SIM.enumerate_attractors">
<span class="sig-name descname"><span class="pre">enumerate_attractors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BONESIS_SIM.enumerate_attractors" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BONESIS_SIM.generate_trajectories">
<span class="sig-name descname"><span class="pre">generate_trajectories</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outputs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BONESIS_SIM.generate_trajectories" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.BONESIS_SIM.initialize_network">
<span class="sig-name descname"><span class="pre">initialize_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.BONESIS_SIM.initialize_network" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MABOSS_SIM">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_sim.</span></span><span class="sig-name descname"><span class="pre">MABOSS_SIM</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seednb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MABOSS_SIM" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#NORDic.UTILS.utils_sim.BN_SIM" title="NORDic.UTILS.utils_sim.BN_SIM"><code class="xref py py-class docutils literal notranslate"><span class="pre">BN_SIM</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MABOSS_SIM.add_initial_states">
<span class="sig-name descname"><span class="pre">add_initial_states</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">initial</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MABOSS_SIM.add_initial_states" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MABOSS_SIM.add_permanent_mutation">
<span class="sig-name descname"><span class="pre">add_permanent_mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MABOSS_SIM.add_permanent_mutation" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MABOSS_SIM.add_transient_mutation">
<span class="sig-name descname"><span class="pre">add_transient_mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MABOSS_SIM.add_transient_mutation" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MABOSS_SIM.enumerate_attractors">
<span class="sig-name descname"><span class="pre">enumerate_attractors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MABOSS_SIM.enumerate_attractors" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MABOSS_SIM.generate_trajectories">
<span class="sig-name descname"><span class="pre">generate_trajectories</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outputs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MABOSS_SIM.generate_trajectories" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MABOSS_SIM.initialize_network">
<span class="sig-name descname"><span class="pre">initialize_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MABOSS_SIM.initialize_network" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MPBN_SIM">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_sim.</span></span><span class="sig-name descname"><span class="pre">MPBN_SIM</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">seednb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MPBN_SIM" title="Permalink to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#NORDic.UTILS.utils_sim.BN_SIM" title="NORDic.UTILS.utils_sim.BN_SIM"><code class="xref py py-class docutils literal notranslate"><span class="pre">BN_SIM</span></code></a></p>
<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MPBN_SIM.add_initial_states">
<span class="sig-name descname"><span class="pre">add_initial_states</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">initial</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MPBN_SIM.add_initial_states" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MPBN_SIM.add_permanent_mutation">
<span class="sig-name descname"><span class="pre">add_permanent_mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MPBN_SIM.add_permanent_mutation" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MPBN_SIM.add_transient_mutation">
<span class="sig-name descname"><span class="pre">add_transient_mutation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mutation</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MPBN_SIM.add_transient_mutation" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MPBN_SIM.enumerate_attractors">
<span class="sig-name descname"><span class="pre">enumerate_attractors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_attrs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">-1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MPBN_SIM.enumerate_attractors" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MPBN_SIM.generate_trajectories">
<span class="sig-name descname"><span class="pre">generate_trajectories</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">params</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">outputs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">show_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MPBN_SIM.generate_trajectories" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.MPBN_SIM.initialize_network">
<span class="sig-name descname"><span class="pre">initialize_network</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">network_fname</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.MPBN_SIM.initialize_network" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.capture">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_sim.</span></span><span class="sig-name descname"><span class="pre">capture</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.capture" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.choice">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_sim.</span></span><span class="sig-name descname"><span class="pre">choice</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">a</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">replace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.choice" title="Permalink to this definition"></a></dt>
<dd><p>Generates a random sample from a given 1-D array</p>
<div class="versionadded">
<p><span class="versionmodified added">New in version 1.7.0.</span></p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>New code should use the <cite>~numpy.random.Generator.choice</cite>
method of a <cite>~numpy.random.Generator</cite> instance instead;
please see the <span class="xref std std-ref">random-quick-start</span>.</p>
</div>
<section id="id125">
<h3>Parameters<a class="headerlink" href="#id125" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>a<span class="classifier">1-D array-like or int</span></dt><dd><p>If an ndarray, a random sample is generated from its elements.
If an int, the random sample is generated as if it were <code class="docutils literal notranslate"><span class="pre">np.arange(a)</span></code></p>
</dd>
<dt>size<span class="classifier">int or tuple of ints, optional</span></dt><dd><p>Output shape.  If the given shape is, e.g., <code class="docutils literal notranslate"><span class="pre">(m,</span> <span class="pre">n,</span> <span class="pre">k)</span></code>, then
<code class="docutils literal notranslate"><span class="pre">m</span> <span class="pre">*</span> <span class="pre">n</span> <span class="pre">*</span> <span class="pre">k</span></code> samples are drawn.  Default is None, in which case a
single value is returned.</p>
</dd>
<dt>replace<span class="classifier">boolean, optional</span></dt><dd><p>Whether the sample is with or without replacement. Default is True,
meaning that a value of <code class="docutils literal notranslate"><span class="pre">a</span></code> can be selected multiple times.</p>
</dd>
<dt>p<span class="classifier">1-D array-like, optional</span></dt><dd><p>The probabilities associated with each entry in a.
If not given, the sample assumes a uniform distribution over all
entries in <code class="docutils literal notranslate"><span class="pre">a</span></code>.</p>
</dd>
</dl>
</section>
<section id="id126">
<h3>Returns<a class="headerlink" href="#id126" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>samples<span class="classifier">single item or ndarray</span></dt><dd><p>The generated random samples</p>
</dd>
</dl>
</section>
<section id="raises">
<h3>Raises<a class="headerlink" href="#raises" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>ValueError</dt><dd><p>If a is an int and less than zero, if a or p are not 1-dimensional,
if a is an array-like of size 0, if p is not a vector of
probabilities, if a and p have different lengths, or if
replace=False and the sample size is greater than the population
size</p>
</dd>
</dl>
</section>
<section id="see-also">
<h3>See Also<a class="headerlink" href="#see-also" title="Permalink to this heading"></a></h3>
<p>randint, shuffle, permutation
random.Generator.choice: which should be used in new code</p>
</section>
<section id="notes">
<h3>Notes<a class="headerlink" href="#notes" title="Permalink to this heading"></a></h3>
<p>Setting user-specified probabilities through <code class="docutils literal notranslate"><span class="pre">p</span></code> uses a more general but less
efficient sampler than the default. The general sampler produces a different sample
than the optimized sampler even if each element of <code class="docutils literal notranslate"><span class="pre">p</span></code> is 1 / len(a).</p>
<p>Sampling random rows from a 2-D array is not possible with this function,
but is possible with <cite>Generator.choice</cite> through its <code class="docutils literal notranslate"><span class="pre">axis</span></code> keyword.</p>
</section>
<section id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this heading"></a></h3>
<p>Generate a uniform random sample from np.arange(5) of size 3:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>
<span class="go">array([0, 3, 4]) # random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#This is equivalent to np.random.randint(0,5,3)</span>
</pre></div>
</div>
<p>Generate a non-uniform random sample from np.arange(5) of size 3:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">array([3, 3, 0]) # random</span>
</pre></div>
</div>
<p>Generate a uniform random sample from np.arange(5) of size 3 without
replacement:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
<span class="go">array([3,1,0]) # random</span>
<span class="gp">&gt;&gt;&gt; </span><span class="c1">#This is equivalent to np.random.permutation(np.arange(5))[:3]</span>
</pre></div>
</div>
<p>Generate a non-uniform random sample from np.arange(5) of size
3 without replacement:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="mi">5</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">,</span> <span class="mf">0.6</span><span class="p">,</span> <span class="mi">0</span><span class="p">])</span>
<span class="go">array([2, 3, 0]) # random</span>
</pre></div>
</div>
<p>Any of the above can be repeated with an arbitrary array-like
instead of just integers. For instance:</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">aa_milne_arr</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;pooh&#39;</span><span class="p">,</span> <span class="s1">&#39;rabbit&#39;</span><span class="p">,</span> <span class="s1">&#39;piglet&#39;</span><span class="p">,</span> <span class="s1">&#39;Christopher&#39;</span><span class="p">]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">aa_milne_arr</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="n">p</span><span class="o">=</span><span class="p">[</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.1</span><span class="p">,</span> <span class="mf">0.3</span><span class="p">])</span>
<span class="go">array([&#39;pooh&#39;, &#39;pooh&#39;, &#39;pooh&#39;, &#39;Christopher&#39;, &#39;piglet&#39;], # random</span>
<span class="go">      dtype=&#39;&lt;U11&#39;)</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_sim.test">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_sim.</span></span><span class="sig-name descname"><span class="pre">test</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">enumerator</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seednb</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">control_profile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">treated_profiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compare_to</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_permanent</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation_transient</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">{}</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">gene_outputs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">print_boxplot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_sim.test" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="module-NORDic.UTILS.utils_state">
<span id="nordic-utils-utils-state-module"></span><h2>NORDic.UTILS.utils_state module<a class="headerlink" href="#module-NORDic.UTILS.utils_state" title="Permalink to this heading"></a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_state.binarize_experiments">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_state.</span></span><span class="sig-name descname"><span class="pre">binarize_experiments</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">thres</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'binary'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_state.binarize_experiments" title="Permalink to this definition"></a></dt>
<dd><p>Binarize experimental profiles</p>
<p>…</p>
<section id="id127">
<h3>Parameters<a class="headerlink" href="#id127" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>data<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[samples]</p>
</dd>
<dt>thres<span class="classifier">Python float</span></dt><dd><p>[default=0.5] : threshold for &#64;method=”binary” (in [0,0.5])</p>
</dd>
<dt>method<span class="classifier">Python character string</span></dt><dd><p>[default=”binary”] : binarization method in [“binary”,”probin”]</p>
</dd>
<dt>strict<span class="classifier">Python bool</span></dt><dd><p>[default=True] : takes into account equalities (if set to True, value=thres will lead to undefined for the corresponding gene)</p>
</dd>
<dt>njobs<span class="classifier">Python integer</span></dt><dd><p>[default=1] : parallelism if needed</p>
</dd>
</dl>
</section>
<section id="id128">
<h3>Returns<a class="headerlink" href="#id128" title="Permalink to this heading"></a></h3>
<p>signatures : Pandas DataFrame: rows/[genes] x columns[samples] with values in [0,1,NaN]</p>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_state.compare_states">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_state.</span></span><span class="sig-name descname"><span class="pre">compare_states</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">x</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_state.compare_states" title="Permalink to this definition"></a></dt>
<dd><p>Computes the similarity between two sets of Boolean states</p>
<p>…</p>
<section id="id129">
<h3>Parameters<a class="headerlink" href="#id129" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>x<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[state IDs] contains (0, 1, NaN)</p>
</dd>
<dt>y<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[state IDs] contains (0, 1, NaN)</p>
</dd>
<dt>genes<span class="classifier">Python character string list</span></dt><dd><p>list of gene symbols</p>
</dd>
</dl>
</section>
<section id="id130">
<h3>Returns<a class="headerlink" href="#id130" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>sims<span class="classifier">NumPy array</span></dt><dd><p>similarities between each column of x and each columns of y, on the list of N present genes in genes (if provided) otherwise on the union of N genes in x and y</p>
</dd>
<dt>N<span class="classifier">Python integer</span></dt><dd><p>number of genes on which the similarity is computed</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_state.finetune_binthres">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_state.</span></span><span class="sig-name descname"><span class="pre">finetune_binthres</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">samples</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">network_fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mutation</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step=0.005</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">maxt=0.5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mint=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_binthres=&lt;function</span> <span class="pre">&lt;lambda&gt;&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs=1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose=True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_state.finetune_binthres" title="Permalink to this definition"></a></dt>
<dd><p>Select the binarization threshold (in function &#64;binarize_experiments) which maximize the dissimilarity interconditions and the similarity intracondition
…</p>
<section id="id131">
<h3>Parameters<a class="headerlink" href="#id131" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>df<span class="classifier">Pandas DataFrame</span></dt><dd><p>rows/[genes] x columns/[samples]: profiles</p>
</dd>
<dt>samples<span class="classifier">Python character string list</span></dt><dd><p>annotations of conditions for each sample in df</p>
</dd>
<dt>network_fname<span class="classifier">Python character string</span></dt><dd><p>file name containing the network</p>
</dd>
<dt>mutation<span class="classifier">Python dictionary</span></dt><dd><p>dictionary (key=gene, value=perturbation type) gene perturbations which are considered</p>
</dd>
<dt>step<span class="classifier">Python float</span></dt><dd><p>[default=0.005] step in the interval to look for the threshold value</p>
</dd>
<dt>maxt<span class="classifier">Python float</span></dt><dd><p>[default=0.5] maximum threshold value</p>
</dd>
<dt>mint<span class="classifier">Python float</span></dt><dd><p>[default=0.] minimum threshold value</p>
</dd>
<dt>score_binthres<span class="classifier">Python lambda function</span></dt><dd><p>[default=lambda itc,ita_c,ita_t:(1-itc)*ita_c*ita_t] fitness function for the threshold value</p>
</dd>
<dt>njobs<span class="classifier">Python integer</span></dt><dd><p>[default=1] number of parallel jobs</p>
</dd>
<dt>verbose<span class="classifier">Python bool</span></dt><dd><p>[default=True] prints out verbose</p>
</dd>
</dl>
</section>
<section id="id132">
<h3>Returns<a class="headerlink" href="#id132" title="Permalink to this heading"></a></h3>
<dl class="simple">
<dt>max_thres<span class="classifier">Python float</span></dt><dd><p>threshold value maximizing the fitness function</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="NORDic.UTILS.utils_state.quantile_normalize">
<span class="sig-prename descclassname"><span class="pre">NORDic.UTILS.utils_state.</span></span><span class="sig-name descname"><span class="pre">quantile_normalize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">njobs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#NORDic.UTILS.utils_state.quantile_normalize" title="Permalink to this definition"></a></dt>
<dd></dd></dl>

</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Clémence Réda, Andrée Delahaye-Duriez.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>